sdk/core/azure-json-reflect/src/main/java/com/azure/json/reflect/GsonJsonReader.java

```diff
@@ -0,0 +1,460 @@
+        nextStringMethod = publicLookup.findVirtual(gsonReaderClass, "nextString", methodType(String.class));
+        nextDoubleMethod = publicLookup.findVirtual(gsonReaderClass, "nextDouble", methodType(double.class));
+        nextIntMethod = publicLookup.findVirtual(gsonReaderClass, "nextInt", methodType(int.class));
+        nextLongMethod = publicLookup.findVirtual(gsonReaderClass, "nextLong", methodType(long.class));
+        nextNameMethod = publicLookup.findVirtual(gsonReaderClass, "nextName", methodType(String.class));
+        skipValueMethod = publicLookup.findVirtual(gsonReaderClass, "skipValue", voidMT);
+
+        initialized = true;
+    }
+
+    @Override
+    public JsonToken currentToken() {
+        return currentToken;
+    }
+
+    @Override
+    public JsonToken nextToken() throws IOException {
+        if (complete) {
+            return currentToken;
+        }
+
+        // GSON requires explicitly beginning and ending arrays and objects and consuming null values.
+        // The contract of JsonReader implicitly overlooks these properties.
+        try {
+            if (currentToken == JsonToken.START_OBJECT) {
+                beginObjectMethod.invoke(gsonReader);
+            } else if (currentToken == JsonToken.END_OBJECT) {
+                endObjectMethod.invoke(gsonReader);
+            } else if (currentToken == JsonToken.START_ARRAY) {
+                beginArrayMethod.invoke(gsonReader);
+            } else if (currentToken == JsonToken.END_ARRAY) {
+                endArrayMethod.invoke(gsonReader);
+            } else if (currentToken == JsonToken.NULL) {
+                nextNullMethod.invoke(gsonReader);
+            }
+
+            if (!consumed && currentToken != null) {
+                switch (currentToken) {
+                    case FIELD_NAME:
+                        nextNameMethod.invoke(gsonReader);
+                        break;
+
+                    case BOOLEAN:
+                        nextBooleanMethod.invoke(gsonReader);
+                        break;
+
+                    case NUMBER:
+                        nextDoubleMethod.invoke(gsonReader);
+                        break;
+
+                    case STRING:
+                        nextStringMethod.invoke(gsonReader);
+                        break;
+
+                    default:
+                        break;
+                }
+            }
+
+            currentToken = mapToken((Enum<?>) peekMethod.invoke(gsonReader));
+
+            if (currentToken == JsonToken.END_DOCUMENT) {
+                complete = true;
+            }
+
+            consumed = false;
+            return currentToken;
+
+        } catch (Throwable e) {
+            if (e instanceof IOException) {
+                throw (IOException) e.getCause();
+            } else {
+                throw (RuntimeException) e.getCause();
+            }
```

Comment:
These catch blocks aren't exactly correct as there are other types of exceptions that could happen, and the cause exception isn't the same as the exception:

```java
if (e instanceof IOException) {
  throw (IOException) e;
} else if (e instanceof RuntimeException) {
  throw (RuntimeException) e;
} else if (e instanceof Error) {
  throw (Error) e;
} else {
  throw new RuntimeException(e);
}
```

This will make sure that all potential `Throwable`s are properly handled. Should also turn this into a static method on this class, or in a package private utility method, as this is repeated often.

Reply:
These catches are being avoiding by using `LambdaMetaFactory` which is nearing completion